drivers/soundwire/intel_master.c

```diff
@@ -0,0 +1,203 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright(c) 2020 Intel Corporation.
+
+#include <linux/device.h>
+#include <linux/acpi.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#if !IS_ENABLED(CONFIG_VIRTUAL_BUS)
+#include <linux/platform_device.h>
+#else
+#include <linux/virtual_bus.h>
+#endif
+#include "cadence_master.h"
+#include "intel_master.h"
+#include "intel.h"
+#include "bus.h"
+
+/*
+ * this driver is an abstraction of the virtbus_device or
+ * platform_device, with additional data structures needed to help
+ * manage the Intel SoundWire IP and completion structure to force a
+ * synchronous behavior so that the bus enumeration is complete before
+ * the parent can continue its initialization. This is needed e.g. to
+ * detect the platform configuration and identify the relevant ASoC
+ * machine driver.
+ */
+#if IS_ENABLED(CONFIG_VIRTUAL_BUS)
+
+static void sdw_intel_master_release(struct virtbus_device *_dev)
+{
+	struct sdw_intel_master_dev *master_dev =
+		to_sdw_intel_master_dev(_dev);
+
+	kfree(master_dev);
+}
+
+#else
+
+static void sdw_intel_master_release(struct device *_dev)
+{
+	struct platform_device *pdev = container_of(_dev, struct platform_device, dev);
+	struct sdw_intel_master_dev *master_dev =
+		to_sdw_intel_master_dev(pdev);
+
+	kfree(master_dev);
+}
+
+#endif
+
+struct sdw_intel_master_dev
+*sdw_intel_master_register(struct device *parent,
+			   struct fwnode_handle *fwnode,
+			   char *name,
+			   int link_id,
+			   struct sdw_intel_link_res *link_res)
+{
+#if IS_ENABLED(CONFIG_VIRTUAL_BUS)
+	struct virtbus_device *vdev;
+#else
+	struct platform_device *pdev;
+#endif
+	struct sdw_intel_master_dev *master_dev;
+
+	unsigned long time, timeout;
+	int ret;
+
+	master_dev = kzalloc(sizeof(*master_dev), GFP_KERNEL);
+	if (!master_dev)
+		return ERR_PTR(-ENOMEM);
+
+	master_dev->link_id = link_id;
+	master_dev->link_res = link_res;
+	master_dev->fwnode = fwnode;
+
+	init_completion(&master_dev->probe_complete);
+
+#if IS_ENABLED(CONFIG_VIRTUAL_BUS)
+	vdev = &master_dev->vdev;
+	vdev->name = devm_kstrdup(parent, name, GFP_KERNEL);
+	if (!vdev->name)
+		return ERR_PTR(-ENOMEM);
+	vdev->release = sdw_intel_master_release;
+	vdev->dev.parent = parent;
+	vdev->dev.fwnode = fwnode;
+	vdev->dev.dma_mask = parent->dma_mask;
+
+	ret = virtbus_register_device(vdev);
+	if (ret < 0)
+		return ERR_PTR(ret);
+#else
+	/*
+	 * we don't use platform_device_alloc() since we want the
+	 * platform device to be encapsulated in the
+	 * sdw_intel_master_device structure.
+	 */
```

Comment:
There's potential danger with this, try playing with bind-unbind and checking whether it still works after that. Also module loading / unloading, eventually runtime-pm. It's safer and easier to play by the book here, which doesn't mean of course, that that's the only way.